replace printing with making the method return values

Jacob-Eliat-Eliat opened the pull request:
Reason: Currently there is a println operation each time we want to initialize the service. 
This is obviously not something we want, especially given that there are "duplicate conversions" as flagged by this method as it stands, and there is AFAIK no plan to remove them.

This PR both removes the println but also changes the method to return a map which can easily be printed somewhere else. This is useful if you need it for any checks later.

polomani commented on src/main/kotlin/com/cognite/units/UnitService.kt:
```diff
@@ -81,28 +81,13 @@ class UnitService(units: String, systems: String) {
-                            conversion.offset +
-                            "] with external IDs:",
-                    )
-                    duplicates.forEach { duplicate ->
-                        println("    ${duplicate.externalId} - ${duplicate.symbol}")
-                    }
```
can this code be somehow run as a part of maven test action?
e.g. so the duplicates are getting printed out in CICD and while running the tests locally

Jacob-Eliat-Eliat replied:
I'll try to figure out a good way to print it in PRs, I still don't think it should live there, or even in the tests really, but I get your usecase. I think it might be possible to create an automatic comment on PRs with changes to the files with the now-deleted print's content in it, but I have to find a good way to call kotlin code from github actions.

evertoncolling replied:
Otherwise we could completely decouple the tests/verifications related to the "content" of the catalog from the tests related to the "library". Most PRs on this repo will be related to content, and once all services read the units dynamically from the API, we won't be needing to redeploy often the maven library either.

Jacob-Eliat-Eliat replied:
I tried this automatic comment idea, it's a bit harder than I though, my approach did not work at all, it's easy to create a comment in github action, what's hard is to somehow load the unitservice, call the method and then format the results. I still think it's possible but it might need some proper kotlin project living in github actions with dependency resolutions and all that stuff.

Jacob-Eliat-Eliat replied:
In the meantime should I just put back the printing but in the test and let your team get back into this later (or let your team handle it)?

evertoncolling replied:
I know it's not ideal, but to allow you to move on fast, one option would be to add an optional argument to the service constructor, like:
```kotlin
class UnitService(units: String, systems: String, checkDuplicates: Boolean = true)
```
which can be propagated and used to control whether to print the duplicates or not. This would allow you and other users of the service to opt out of the printing, and we avoid having to make any changes on the GitHub actions right now.

Later in the future, we could plan to work on the GitHub actions and solve this in a better way.

polomani replied:
Agree with Everton, it's not beautiful but it'll resolve the issue.
Another way is to use an actual Logger with level DEBUG instead of the printing directly, that way you won't get this printed in the prod runtime (though it still runs the code, so it potentially wastes a few ms)

Jacob-Eliat-Eliat commented:
Given your feedback, I have: 
 * Moved the print to the test
 * Changed the pipeline to run the specific tests so the result will be output

In addition I have also created an artefact containing the duplicated in the pipeline.

In it's something we need, I can suggest:
 * running the pipeline again but this time on the master
 * computing the diff between the two pipelines
 * adding a comment to the PR if there is any diff

But for now, this should replace and improve the current workflow and I suggest this be merged as a first step.

Jacob-Eliat-Eliat commented:
I decided to add the comment directly from the pipeline. This should make checking for duplicates easier!

Jacob-Eliat-Eliat commented:
Given the feedback I have modified the pipeline to show the duplicates as a comment to the PR. It is further possible to show only the changes between PRS, which I think would be an improvement, but this already helps a lot. 

It also allowed me to see that my code to find the duplicates was flagging all units as duplicates... so it's already useful!

Don't pay attention to the inane amount of commit, I've been playing around with github actions to make this work :)

evertoncolling commented:
> Given the feedback I have modified the pipeline to show the duplicates as a comment to the PR. It is further possible to show only the changes between PRS, which I think would be an improvement, but this already helps a lot.
> 
> It also allowed me to see that my code to find the duplicates was flagging all units as duplicates... so it's already useful!
> 
> Don't pay attention to the inane amount of commit, I've been playing around with github actions to make this work :)

Having it as a comment in the PR is much better indeed 🤩. I'll make a commit later to improve the message (making it clear what a reviewer should check on the duplicates).

Jacob-Eliat-Eliat commented:
> > Given the feedback I have modified the pipeline to show the duplicates as a comment to the PR. It is further possible to show only the changes between PRS, which I think would be an improvement, but this already helps a lot.
> > It also allowed me to see that my code to find the duplicates was flagging all units as duplicates... so it's already useful!
> > Don't pay attention to the inane amount of commit, I've been playing around with github actions to make this work :)
> 
> Having it as a comment in the PR is much better indeed 🤩. I'll make a commit later to improve the message (making it clear what a reviewer should check on the duplicates).

Yeah it's also possible to checkout the main branch, then run the same test in the main branch, and compare the outputs, then make a comment with the diff in instead of the entire list, but I though I should keep it simpler for this PR.

There are some downsides too, it creates a comment each time the pipeline runs, but it's easy to delete the extra comments.

polomani reviewed (approved):
LGTM
